### PR TITLE
More Memory-efficient Implementation of itertools.cycle

### PR DESCRIPTION
**Pull request by youkaichao**
Requires no additional copy of the iterable.

Standard implementation of ``cycle`` looks like:
```
def cycle(iterable): # v1
    # cycle('ABCD') --> A B C D A B C D A B C D ...
    saved = []
    for element in iterable:
        yield element
        saved.append(element)
    while saved:
        for element in saved:
              yield element
```

We propose to change it to 
```
def cycle(iterable): # v2
    # cycle('ABCD') --> A B C D A B C D A B C D ...
    while True:
        for element in iterable:
              yield element
```

## Drawback: a little incompatible with the previous ``cycle``

v1 can accept an iteratable which can be iterated only once while v2 cannot.

## Benefit: more memory-efficient 

It doesn't require an additional copy.

I think v2 is better. If the caller wants to ``cycle`` through an iteratable (named ``it``) which can be iterated only once, he can use ``cycle(list(it))``. In most cases, the caller holds repeatable iterable, so ``cycle(it)`` is ok without additional memory consumption.

## Benefit for DeepLearning

And there are cases where **multiple iter calls to an iterable returns different iterators!** The [dataloader](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader) for DeepLearning is a perfect example:
```
# dl is a dataloader
first = [img, label for img, label in dl]
second = [img, label for img, label in dl]
assert first[0] != second[0]
```

What's more, in DeepLearning, it is not possible to store the whole dataset in memory, but the dataloader can be called multiple times to get the iterator:
```
# v1
cycle(dl) # memory not enough

# v2
cycle(dl) # no problem
```

So I hope we can get this more memory efficient ``cycle`` :)

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@youkaichao

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by asvetlov** [comment]
Sorry, the incompatibility is the reason for the PR rejection; the behavior change breaks a lot of existing code.
Your implementation may exist under a different name but before proposing to add this into stdlib please publish it on PyPI as a third-party library.
After battle testing and getting a positive feedback you can consider adding it into the standard `itertools` module.

**Comment by youkaichao** [comment]
Hi, does python can tell if an ``iterable`` can be iterated multiple times? If yes, we can enjoy the memory-efficient ``cycle``.

**Comment by tim-one** [comment]
> Hi, does python can tell if an `iterable` can be iterated multiple times? If yes, we can enjoy the memory-efficient `cycle`.

If Python supported that, `cycle` would already have been written to exploit that :wink:.  To be very clear, no, there is no way to tell in advance whether an iterable is "reiterable".

**Comment by youkaichao** [comment]
How obout this:

```
def can_reiterable(x):
    return hasattr(x, '__iter__') and not hasattr(x, '__next__')
```

The hierarchy looks like:
Iterable
|_Iterator

Iterators can be iterated only once, but an Iterable object without being an iterator can be iterated multiple times.

**Comment by youkaichao** [comment]
And how about this:

```
def iterated_once(x):
    return iter(x) is iter(x)
```

**If an iterable returns the same iterator, then it means it can be iterated only once.**

**Comment by youkaichao** [comment]
I opened a question in StackOverflow, getting a similar [answer](https://stackoverflow.com/a/59572849/9191338). It seems by comparing multiple returned iterators, we can tell if the iterable can be iterated multiple times :)

**Comment by asvetlov** [comment]
Very often it is true but the language doesn't guarantee the re-iterability if `iter()` always return a new object.

**Comment by youkaichao** [comment]
@asvetlov If ``iter()`` always returns a new object, but each returned iterator contains different sequence, then what's the semantic meaning of ``cycle``?

Check this:
```
class Strange(object):
    def __init__(self):
        self.i = 0

    def __iter__(self):
        self.i +=2
        for i in range(self.i, self.i + 2):
            yield i
```

Should ``cycle(Strange())`` return ``[2,3, 4, 5, 6, ...]`` or ``[2, 3, 2, 3, 2, 3, 2, 3, ...]``?

**Comment by asvetlov** [comment]
As @tim-one said, there is no reliable way to detect if an iterable is re-iterable or not. Period.
Given this, `cicle()` cannot be optimized. Period.
Sorry, there is no space for improvements here.

**Comment by youkaichao** [comment]
Got that. Now I just wonder how to correctly understand ``cycle(Strange())``?

**Comment by asvetlov** [comment]
`list(Strange()) == [2, 3]`.
I guess `cycle(Strange())` should produce `2, 3, 2, 3, 2, 3, ...` infinite sequence.

**Comment by youkaichao** [comment]
Okay. Although it looks strange to me, I have to accept that. Personally I prefer ``cycle(Strange())`` to produce ``[2,3, 4, 5, 6, ...]``, as in the dataloader example for deeplearning.

**Comment by sthagen** [comment]
The increment of the __iter__ method is outside of the for loop of the yield expression, so how could it be reached again without calling __iter__() explicitly more than once?

**Comment by youkaichao** [comment]
@sthagen  I cannot get your answer. ``i`` is a local variable, while ``self.i`` is a member. Does it answer your question?

**Comment by tim-one** [comment]
I just happened to write a class for doing topological sorts (a graph algorithm) this week.  Since the purpose is to impose a linear order on a graph's nodes, a natural interface was to give it an `__iter__()` method to generate the nodes in that order.

So `iter(instance)` is different each time it's called, and different from `instance` too, but actually iterating more than once would give insane results:  internal info is mutated as a side effect of _computing_ the linear order.  As is, trying would cause an `assert` to fail (a node's predecessor count would fall below 0, which should be impossible) - but without that `assert` it could go on to generate an order that had nothing to do with the (correct!) order generated the first time.

While it's hard to imagine why I would pass an instance of that to `cycle()` :wink:, if I did the only sensible thing `cycle()` could do is capture the first iteration's (correct!) results and replay them.

**Comment by youkaichao** [comment]
So there are two possible semantic meaning of ``cycle``. 
```
class Strange(object):
    def __init__(self):
        self.i = 0

    def __iter__(self):
        self.i +=2
        for i in range(self.i, self.i + 2):
            yield i
```
As pointed in my previous comment, ``cycle(Strange())`` can eturn ``[2,3, 4, 5, 6, ...]`` or ``[2, 3, 2, 3, 2, 3, 2, 3, ...]``, depending on the specific requirements. 

@tim-one  gives an example when we must return ``[2, 3, 2, 3, 2, 3, 2, 3, ...]``, while I give an example when we must return ``[2,3, 4, 5, 6, ...]``: if the object is a dataset loader, each call ``iter(obj)`` returns a **differently** shuffled iterator of the dataset, then we have to return  ``[2,3, 4, 5, 6, ...]`` in order to train a neural network.

It seems better to split these two semantic meanings into two functions:
```
def cycle1(iterable):
    saved = []
    for element in iterable:
        yield element
        saved.append(element)
    while saved:
        for element in saved:
              yield element

def cycle2(iterable):
    while True:
        for element in iterable:
              yield element
```

**Comment by tim-one** [comment]
Seems to me that what you want can be gotten easily and efficiently today via:

```python
    from itertools import repeat, chain
    def cycle2(iterable):
        return chain.from_iterable(repeat(iterable))
```

Perhaps that could be added as a recipe.  The name "cycle" really doesn't fit, though.  What's being cycled _through_?  As the implementation above makes clear, the only thing being visited repeatedly is the iterable itself, and "cycle of length 1" is a strained way to say "repeat" ;-)

**Comment by youkaichao** [comment]
@tim-one  Yes, you are right.  But I just found one bug: if the input is a ``generator``, say: ``a = cycle2(i for i in range(1))``, then the second call to ``next(a)`` will stuck. I think the semantic meaning of ``cycle2`` is to **concate multiple iterators from an iterable**.
